watchdog.go

```diff
@@ -117,43 +122,57 @@ func Memory(config MemConfig) (err error, stop func()) {
 		config.Limit = mem.Total
 	}
 
-	watchdog.config = config
-	watchdog.closing = make(chan struct{})
+	_watchdog.config = config
+	_watchdog.closing = make(chan struct{})
 
-	watchdog.wg.Add(1)
+	_watchdog.wg.Add(1)
 	go watch()
 
 	return nil, stopMemory
 }
 
 func watch() {
 	var (
-		lk          sync.Mutex // guards gcTriggered channel, which is drained and flipped to nil on closure.
 		gcTriggered = make(chan struct{}, 16)
 
 		memstats runtime.MemStats
 		sysmem   gosigar.Mem
-		config   = watchdog.config
+		config   = _watchdog.config
 	)
 
 	// this non-zero sized struct is used as a sentinel to detect when a GC
 	// run has finished, by setting and resetting a finalizer on it.
+	// it essentially creates a GC notification "flywheel"; every GC will
+	// trigger this finalizer, which will reset itself so it gets notified
+	// of the next GC, breaking the cycle when the watchdog is stopped.
 	type sentinel struct{ a *int }
-	var sentinelObj sentinel
 	var finalizer func(o *sentinel)
 	finalizer = func(o *sentinel) {
-		lk.Lock()
-		defer lk.Unlock()
+		if atomic.LoadInt32(&_watchdog.state) != stateRunning {
+			// this GC triggered after the watchdog was stopped; ignore
+			// and do not reset the finalizer.
+			return
+		}
+
+		// reset so it triggers on the next GC.
+		runtime.SetFinalizer(o, finalizer)
+
+		if atomic.LoadInt32(&_watchdog.finalizerSkip) == 1 {
```

Comment:
Nit: we can just:

1. Make the GC trigger buffer length 1.
2. Drain the GC trigger buffer after an explicit GC.

Reply:
Yep, that would work too.

Reply:
Actually, I don't think that approach is safe. We don't control the Go GC pacer. If it decides to run at the same time, we won't know if we're consuming our own signal, or the legitimate Go GC pacer signal. It's way safer if we don't send at all, instead of sending and compensating for that send. That way signals don't get mixed up.

Reply:
I see. Although I'm not sure if that really matters. If we GC twice in a row, we might as well treat it as one GC.